kube-runtime/src/finalizer.rs

```diff
@@ -0,0 +1,146 @@
+///
+/// # Errors
+///
+/// `apply` and `cleanup` are both fallible, their errors are passed through as [`Error::ApplyFailed`]
+/// and [`Error::CleanupFailed`], respectively.
+///
+/// In addition, adding and removing the finalizer itself may fail. In particular, this may be because of
+/// network errors, or because another `finalizer` was updated in the meantime on the same object.
+pub async fn finalizer<K, ApplyFut, CleanupFut>(
+    api: &Api<K>,
+    finalizer_name: &str,
+    obj: K,
+    apply: impl FnOnce(K) -> ApplyFut,
+    cleanup: impl FnOnce(K) -> CleanupFut,
+) -> Result<ReconcilerAction, Error<ApplyFut::Error, CleanupFut::Error>>
+where
+    K: Metadata<Ty = ObjectMeta> + Clone + DeserializeOwned + Serialize + Debug,
+    ApplyFut: TryFuture<Ok = ReconcilerAction>,
+    ApplyFut::Error: StdError + 'static,
+    CleanupFut: TryFuture<Ok = ()>,
+    CleanupFut::Error: StdError + 'static,
+{
+    if let Some((finalizer_i, _)) = obj
+        .metadata()
+        .finalizers
+        .as_ref()
+        .and_then(|fins| fins.iter().enumerate().find(|(_, fin)| *fin == finalizer_name))
+    {
+        if obj.metadata().deletion_timestamp.is_none() {
+            apply(obj).into_future().await.context(ApplyFailed)
+        } else {
+            let name = obj.metadata().name.clone().unwrap();
+            cleanup(obj).into_future().await.context(CleanupFailed)?;
+            let finalizer_path = format!("/metadata/finalizers/{}", finalizer_i);
+            api.patch::<K>(
+                &name,
+                &PatchParams::default(),
+                &Patch::Json(json_patch::Patch(vec![
+                    PatchOperation::Test(TestOperation {
+                        path: finalizer_path.clone(),
+                        value: finalizer_name.into(),
+                    }),
+                    PatchOperation::Remove(RemoveOperation { path: finalizer_path }),
+                ])),
+            )
+            .await
+            .context(RemoveFinalizer)?;
+            Ok(ReconcilerAction { requeue_after: None })
+        }
+    } else {
+        if obj.metadata().deletion_timestamp.is_none() {
+            let patch = json_patch::Patch(if obj.metadata().finalizers.is_none() {
+                vec![
+                    PatchOperation::Test(TestOperation {
+                        path: "/metadata/finalizers".to_string(),
+                        value: serde_json::Value::Null,
+                    }),
+                    PatchOperation::Add(AddOperation {
+                        path: "/metadata/finalizers".to_string(),
+                        value: vec![finalizer_name].into(),
+                    }),
+                ]
+            } else {
+                vec![PatchOperation::Add(AddOperation {
+                    path: "/metadata/finalizers/-".to_string(),
+                    value: finalizer_name.into(),
+                })]
+            });
```

Comment:
Woah. What does TestOperation and AddOperation do together here?
And what is the magic `-`?

Reply:
> What does TestOperation and AddOperation do together here?

Fail the operation if `/metadata/finalizers` was set in the meantime (so we'd rather fail than remove others' finalizers).

> And what is the magic -?

[End of the list](http://jsonpatch.com/#json-pointer)

Reply:
Ahh:

> To point to the root of the document use an empty string for the pointer. The pointer / doesn’t point to the root, it points to a key of "" on the root (which is totally valid in JSON).